packages/images-plugin/lib/compare.js

```diff
@@ -8,76 +8,93 @@ const MATCH_TEMPLATE_MODE = 'matchTemplate';
+ */
+function isMatchTemplateMode(value) {
+  return _.toLower(value) === MATCH_TEMPLATE_MODE.toLowerCase();
+}
+
 /**
  * Performs images comparison using OpenCV framework features.
  * It is expected that both OpenCV framework and opencv4nodejs
  * module are installed on the machine where Appium server is running.
  *
- * @param {string} mode - One of possible comparison modes:
+ * @template {CompareMode} Mode
+ * @template {boolean} Multiple
+ * @param {Mode} mode - One of possible comparison modes:
  * matchFeatures, getSimilarity, matchTemplate
  * @param {string} firstImage - Base64-encoded image file.
  * All image formats, that OpenCV library itself accepts, are supported.
  * @param {string} secondImage - Base64-encoded image file.
  * All image formats, that OpenCV library itself accepts, are supported.
- * @param {object} [options] - The content of this dictionary depends
+ * @param {CompareImagesOptions<Mode,Multiple>} [options] - The content of this dictionary depends
  * on the actual `mode` value. See the documentation on `@appium/support`
  * module for more details.
- * @returns {Promise<object|object[]>} The content of the resulting dictionary depends
+ * @returns {Promise<CompareImagesResult<Mode,Multiple>>} The content of the resulting dictionary depends
  * on the actual `mode` and `options` values. See the documentation on
  * `@appium/support` module for more details.
  * @throws {Error} If required OpenCV modules are not installed or
  * if `mode` value is incorrect or if there was an unexpected issue while
  * matching the images.
  */
-async function compareImages(mode, firstImage, secondImage, options = {}) {
+async function compareImages(
+  mode,
+  firstImage,
+  secondImage,
+  options = /** @type {CompareImagesOptions<Mode,Multiple>} */ ({})
+) {
   const img1 = Buffer.from(firstImage, 'base64');
   const img2 = Buffer.from(secondImage, 'base64');
-  let result = null;
-  switch (_.toLower(mode)) {
-    case MATCH_FEATURES_MODE.toLowerCase():
-      try {
-        result = await getImagesMatches(img1, img2, options);
-      } catch (err) {
-        // might throw if no matches
-        result = {count: 0};
-      }
-      break;
-    case GET_SIMILARITY_MODE.toLowerCase():
-      result = await getImagesSimilarity(img1, img2, options);
-      break;
-    case MATCH_TEMPLATE_MODE.toLowerCase():
-      // firstImage/img1 is the full image and secondImage/img2 is the partial one
-      result = await getImageOccurrence(img1, img2, options);
-      if (options.multiple) {
-        return result.multiple.map(convertVisualizationToBase64);
-      }
-      break;
-    default:
-      throw new errors.InvalidArgumentError(
-        `'${mode}' images comparison mode is unknown. ` +
-          `Only ${JSON.stringify([
-            MATCH_FEATURES_MODE,
-            GET_SIMILARITY_MODE,
-            MATCH_TEMPLATE_MODE,
-          ])} modes are supported.`
-      );
+
+  if (isMatchFeaturesMode(mode)) {
```

Comment:
why switch does not work here?

Reply:
it's a typescript thing--a type guard.  they have to be functions which return a type of `{argName is SomeType}` and you can't do that with a switch, afaik.